docs/proposal/kcl-checksum-verification-checker-module-design.md

```diff
@@ -0,0 +1,173 @@
+# Checker module design for kcl dependency
+```Go
+package Checker
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	pkg "kcl-lang.io/kpm/pkg/package"
+)
+
+type Version struct {
+	major string
+	minor string
+	patch string
+}
+
+func isValidDependency(d pkg.Dependency, localCachePath string) (bool, error) {
+	if !isValidDependencyName(d.Name) {
+		return false, fmt.Errorf("Invalid Dependency Name: %s", d.Name)
+	}
+
+	if !isValidDependencyVersion(d.Version) {
+		return false, fmt.Errorf("Invalid Dependency Version %s for ", d.Version, d.Name)
+	}
+
+	if err := isValidDependencyChecksum(d, localCachePath); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// isValidDependencyName reports whether name of the dependency is appropriate.
+func isValidDependencyName(name string) bool {
+	validNamePattern := `^[a-zA-Z][a-zA-Z_\-\.]*[a-zA-Z]$`
+
+	regex := regexp.MustCompile(validNamePattern)
+
+	return regex.MatchString(name)
+}
+
+// isValidDependencyVersion reports whether v is a valid semantic version string.
+func isValidDependencyVersion(v string) bool {
+	_, ok := parseVersion(v)
+	return ok
+}
+
+// parseVersion parses a semantic version string and returns its major, minor, and patch components.
+func parseVersion(v string) (p Version, ok bool) {
+	p.major, v, ok = parseInt(v)
+	if !ok {
+		return
+	}
+	if v == "" {
+		p.minor = "0"
+		p.patch = "0"
+		return
+	}
+	if v[0] != '.' {
+		ok = false
+		return
+	}
+	p.minor, v, ok = parseInt(v[1:])
+	if !ok {
+		return
+	}
+	if v == "" {
+		p.patch = "0"
+		return
+	}
+	if v[0] != '.' {
+		ok = false
+		return
+	}
+	p.patch, v, ok = parseInt(v[1:])
+	if !ok {
+		return
+	}
+	if len(v) > 0 {
+		ok = false
+	}
+	return
+}
+
+// parseInt extracts an integer from the beginning of a string and returns the integer and the remaining string.
+func parseInt(v string) (t, rest string, ok bool) {
+	if v == "" {
+		return
+	}
+	if v[0] < '0' || '9' < v[0] {
+		return
+	}
+	i := 1
+	for i < len(v) && '0' <= v[i] && v[i] <= '9' {
+		i++
+	}
+	if v[0] == '0' && i != 1 {
+		return
+	}
+	return v[:i], v[i:], true
+}
+
+// HashLocalCache calculates the hash of all files in the specified local cache directory.
+func HashLocalCache(localCachePath string) (string, error) {
```

Comment:
This method has been defined before:

https://github.com/kcl-lang/kpm/blob/bea45753c66734d7f5ebe2a44fe544d14402db13/pkg/utils/utils.go#L32

Reply:
I took inspiration from Go's checksum generation (hence the "h1:" format). Though the differences are subtle, here are the trade-offs and when each method is preferable:
- HashLocalCache: Hashes both file content and name. Use when file names matter.
- HashDir: Can ignore certain files (e.g., .git). Use when excluding non-essential files is necessary.
- HashLocalCache: Slower due to sorting and name hashing. Use for higher accuracy with filenames and ordering.
- HashDir: Faster and simpler. Use when performance is critical and file names/order don’t matter.